pkg/providers/snow/defaults.go

```diff
@@ -25,49 +26,35 @@ func NewDefaulters(aws aws.Clients, writer filewriter.FileWriter) *Defaulters {
 func NewDefaultersFromAwsClientMap(awsClientMap AwsClientMap, writer filewriter.FileWriter) *Defaulters {
 	return &Defaulters{
 		awsClientMap: awsClientMap,
+		writer:       writer,
 	}
 }
 
-func (d *Defaulters) GenerateDefaultSshKey(ctx context.Context, m *v1alpha1.SnowMachineConfig) (keyName, keyVal string, err error) {
+func (d *Defaulters) GenerateDefaultSshKey(ctx context.Context, m *v1alpha1.SnowMachineConfig) error {
 	if m.Spec.SshKeyName != "" {
-		return "", "", nil
-	}
-
-	keyName = fmt.Sprintf("%s-%s.pem", defaultAwsSshKeyNamePrefix, m.GetName())
-
-	logger.V(1).Info("SnowMachineConfig SshKey is empty. Creating default key pair", "default key name", keyName)
-
-	// Only need to fetch IP from single device to create ssh key pair. The created key will then be validated in validateSshKeyPair method
-	var anyDeviceIP string
-	for ip := range d.awsClientMap {
-		anyDeviceIP = ip
-		break
-	}
-	keyVal, err = d.awsClientMap[anyDeviceIP].EC2CreateKeyPair(ctx, keyName)
-	if err != nil {
-		return "", "", fmt.Errorf("creating ssh key pair in snow: %v", err)
+		return nil
 	}
 
-	return keyName, keyVal, nil
-}
-
-func (d *Defaulters) GenerateDefaultSshKeyAndSaveToFile(ctx context.Context, m *v1alpha1.SnowMachineConfig) error {
-	if m.Spec.SshKeyName != "" {
+	if common.ValidateSSHAuthKeyExists(d.writer) {
```

Comment:
another way is to add a field in defaulters struct indicating if the default key is already be generated once and skip the rest.